### ISSUE TITLE
Some Suggestions 

### ISSUE THREAD

**Issue by eliottmusk** [monyxie/qr-lite#36]
Please add:

1 - Dark Theme

2 - A way to delete all history and to delete each history item separately

3 - A setting to disable the history completely, so to not store any history

4 - Options to copy or save the generated QR code

5 - Scan QR in Image never work to me, the plugin need be more flexible, and search for QRs in images with more objects and people etc.

6 - Add a Quick Region Scan, so if you use Alt + Left Click the plugin scan the cursor region and open the URL in a new background (or optionally foreground) tab without any popup, confirmation, just click and open, this Quick mode is to use QRs as a type of link.

7 - Better performance for white codes in black backgrounds

Thanks!

**Comment by monyxie** [comment]
Thanks for the suggestions.

> 1 - Dark Theme

I thought about it in the past. I decided to give up the idea for that the QR code needs a near-white background to achieve maximum readability and that would defeat the purpose of a dark theme.

> 2 - A way to delete all history and to delete each history item separately

This is a good idea. I'll see what I can do.

> 3 - A setting to disable the history completely, so to not store any history

Also a good idea.

> 4 - Options to copy or save the generated QR code

You can already do that via the "Save" and "Copy" buttons right below the generated QR code.

> 5 - Scan QR in Image never work to me, the plugin need be more flexible, and search for QRs in images with more objects and people etc.

The QR code reader has a high chance of failing when trying to reading very small QR codes in large images. This is a limitation of the underlying library so it's unlikely that we will see any changes. You'll have to use the "Select region to scan" feature to scan these kind of images.

"Scan QR in Image" feature could also fail if the extension is not allowed to read the image data. Please provide more detail so I can determine if this is the case.

> 6 - Add a Quick Region Scan, so if you use Alt + Left Click the plugin scan the cursor region and open the URL in a new tab without any popup, confirmation, just click and open, this Quick mode is to use QRs as a type of link.

This sounds like an interesting idea. However listening for mouse clicks would require injecting a script into every single web page the user visits, which is something I'm actively trying to avoid. Additionally, altering left click behavior (even with a modifier key) would likely cause problems on websites that already do that themselves.

For the above reasons I don't think this feature is worth implementing.

> 7 - Better performance for white codes in black backgrounds

This is also a limitation of the underlying library. However I'll see if I can do something to improve sucess rate on my side.

**Comment by eliottmusk** [comment]
First thanks your attention and answers, about dark theme and if you keep white background just on QR code and make the QR small when in Dark Theme? About Save and Copy buttons I can't see both, but I think is because I use "layout.css.devPixelsPerPx: 1.4". About Quick Region Scan, some alternatives would be keep this feature optional and turned off by default, so if the user don't enable it the plugin don't inject anything, and also add a blacklist to exclude sites that has some conflict like "Yet Another Smooth Scrolling WE" plugin. Thank you!

**Comment by eliottmusk** [comment]
alt + right click = open qr code link in current tab

alt + shift + right click = open in a new background tab

**Comment by Mike-Reddick** [comment]
I agree with all this features requests and the Quick Region Scan is really  awesome!

**Comment by monyxie** [comment]
I just released version 3.1.0, which addressed _1_, _2_, _3_, _5_ and _7_. _4_ was already possible before this.
Unfortunely, for now, I really don't think the quick region scan idea is worth implementing for the following reasons: 
- Requires injecting a script into every page. Bear with me when I say this, but I simply don't like the idea, that a QR Code scanner injects a script into every page the user visits. I wouldn't use an QR Code scanner that does this.
- Requires `<all_urls>` permission. Although the extension already depends on it at the present, the extension API is moving torwards making more and more things possible without it. So I'm still hoping I could remove it someday and I don't really like the idea of adding another feature that depends on it.
- Could break websites. Plus I don't want to ship a feature that is disabled by default, and I don't have the time or energy to maintain a blacklist.

**Comment by eliottmusk** [comment]
Thanks so much for the changes, congrats! I want add some notes:

1 - Dark Theme: When you scan with "Region to Scan" the plugin show a popup with the link and available actions, this popup continue using the light theme need be updated.

6 - Quick Region Scan: I understand your position, another option can be implement 3 new keyboard shortcuts to open link directly, this way no need any new script or permission and the user have 4 shortcuts for scan region, one with link review popup and 3 without popup. Example of the keyboard shortcuts:

1 - Select region to scan (with link review popup)

2 - Select Region to Scan (open link directly in current tab)

3 - Select Region to Scan (open link directly in a new background tab)

4 - Select Region to Scan (open link directly in a new foreground tab)

Using the options 2, 3 'n 4 to open directly, if the QR Code is not a link, for instance is a simple text, in this case show the popup with the content anyway.

**Comment by monyxie** [comment]
> 1 - Dark Theme: When you scan with "Region to Scan" the plugin show a popup with the link and available actions, this popup continue using the light theme need be updated.

This will be addressed in the next release.

> 6 - Quick Region Scan: I understand your position, another option can be implement 3 new keyboard shortcuts to open link directly, this way no need any new script or permission and the user have 4 shortcuts for scan region, one with link review popup and 3 without popup. Example of the keyboard shortcuts:
1 - Select region to scan (with link review popup)
2 - Select Region to Scan (open link directly in current tab)
3 - Select Region to Scan (open link directly in a new background tab)
4 - Select Region to Scan (open link directly in a new foreground tab)
Using the options 2, 3 'n 4 to open directly, if the QR Code is not a link, for instance is a simple text, in this case show the popup with the content anyway.

I like this idea better. It'll likely also be in the next release if nothing goes wrong.

**Comment by eliottmusk** [comment]
Nice, I want to suggest another things to you think about, is just some ideas to you:

1 - Setting to enable Auto Scan mode when using Region Scan, this way no need click to scan, when the plugin recognize the QR on square, the plugin scan automatically, the same way that Android scanners do.

2 - A new scan mode and keyboard shortcut to Scan All Visible Region, like the Save Visible of Take Screen Shot feature of Firefox but used to scan QR. This can be interesting for instance when you is watching a youtube video with a QR, so you can simple use the new shortcut and the plugin scan all visible area and show the review popup.

3 - On link review popup add a new button/action to "Open as URL", because has some QR codes that do not include https:// so the plugin do not show Open URL action.

**Comment by monyxie** [comment]
> 1 - Setting to enable Auto Scan mode when using Region Scan, this way no need click to scan, when the plugin recognize the QR on square, the plugin scan automatically, the same way that Android scanners do.

"Select scan region" mode needs to take a partial screenshot of the page. This is a pretty slow operation and could take as long as a second to complete. For a smooth user experience it would need to be much faster than that.

> 2 - A new scan mode and keyboard shortcut to Scan All Visible Region, like the Save Visible of Take Screen Shot feature of Firefox but used to scan QR. This can be interesting for instance when you is watching a youtube video with a QR, so you can simple use the new shortcut and the plugin scan all visible area and show the review popup.

This will often create a situation where the QR code occupy a very small portion of the entire image. Based on my own experience, the QR code reader's detection rate is significantly worse in such situation. 
Capturing the entire screen for scanning is also slow.
Needs more experimenting to decide if the performance is good enough.

> 3 - On link review popup add a new button/action to "Open as URL", because has some QR codes that do not include https:// so the plugin do not show Open URL action.

I don't want to make it possible to open *everything* as URL. However it's worth considering to make things that looks like a partial URL and acceptable by the browser to be openable as URL. I need to find a way to do that first.

**Comment by monyxie** [comment]
> 2 - A new scan mode and keyboard shortcut to Scan All Visible Region, like the Save Visible of Take Screen Shot feature of Firefox but used to scan QR. This can be interesting for instance when you is watching a youtube video with a QR, so you can simple use the new shortcut and the plugin scan all visible area and show the review popup.

After some testing, unfortunately it does seem the scanner is very likely to fail in this kind of situation. For example, it fails to detect the otherwise perfectly scannable QR codes in the following image.
![Image](https://github.com/user-attachments/assets/15f66f0d-da9e-467e-b919-afdd0fcdedb9)

Another possible way to approach this is to iterate through every single visible image or video on the page and try to scan each one of them. But this could also fail in some situations e.g. the image is displayed as CSS background.

**Comment by eliottmusk** [comment]
Thank you for your attention and answers, I think your positions is reasonable, but who knows maybe someday the technology can make possible the implementation of some of this ideas. Also I want to say that after v3.1.0 the extenssion has much better scanner performance, congrats!
